datastore/longhorn.go

```diff
@@ -197,44 +197,55 @@
~	return nil
~}
~
 func (s *DataStore) syncSettingCRsWithCustomizedDefaultSettings(customizedDefaultSettings map[string]string, defaultSettingCMResourceVersion string) error {
 	for _, sName := range types.SettingNameList {
 		configMapResourceVersion := ""
-		if s, err := s.GetSettingExactRO(sName); err != nil {
+		setting, err := s.GetSettingExactRO(sName)
+		if err != nil {
 			if !apierrors.IsNotFound(err) {
 				return err
 			}
 		} else {
-			if s.Annotations != nil {
-				configMapResourceVersion = s.Annotations[types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey)]
+			if setting.Annotations != nil {
+				configMapResourceVersion = setting.Annotations[types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey)]
+			}
+		}
+
+		if sName == types.SettingNamePriorityClass {
+			volumesDetached, err := s.AreAllVolumesDetachedState()
+			if err != nil {
+				return errors.Wrapf(err, "failed to list volumes to check if all volumes are detached")
+			}
+			if !volumesDetached || (setting != nil && setting.Value != "") {
+				continue
```

Comment:
Can we create a warning log to inform users that the setting will not be updated because there are volumes attached?

I assume the controller will not trigger this function except when there is a new change from the config map of Longhorn's default settings. Please confirm this to prevent any unexpected log flooding.

Reply:
Yes, it would be triggered only when the daemon is starting or the config map of Longhorn's default settings has been changed.
Modified. Thanks